development-practices/defining-the-boundaries-of-a-package.md

```diff
@@ -0,0 +1,107 @@
+
+> **The granule of reuse is the granule of release. Only components that are released through a tracking system can be effectively reused. This granule is the package **
+
+Reusability is one of the most oft claimed goals of OOD. But what is reuse? &#x20;
+
+Reuse can come in different flavours
+
+* Textually inserting a bunch of code/components from one source to another
+* Utilizing an existing module from someone else and brining into the current source
+
+The above examples of Reuse are of code/component copying; and it comes with a serious disadvantage, one must own the code/component that we copied. If the code/component doesn't work in the target environment, it must be changed. If there are bugs, it might be fixed. Fixes from the original author must be manually merged into the source.  Eventually the code/component copied diverges from the original that it can be hardly. While code/component copying can make it easier to do some initial development; it does not help very much with the most expensive phase of the software lifecycle, _**maintenance**_.
+
+REP states that the granule of reuse can be no smaller than the granule of release. Anything that we reuse must also be released. Packages are a candidate for a releasable entity. It might be possible to release and track classes or components, but there are so many components in a typical salesforce org, that this would certainly overwhelm the release tracking system. We need some larger scale entity to act as the granule of release; and the package fits this need well.
+
+#### The Common Closure Principle (CCP)
+
+> **The components in a package are reused together. If you reuse one of the components in a package, you reuse them all.**
+
+This principle helps us to decide which components should be placed into a package. It states that components \[1] that tend to be reused together belong in the same package. Components are seldom reused in isolation. Reusable classes/components collaborate with other classes/components that are part of the reusable abstraction. The **CRP** states that these classes/components belong together in the same package.
+
+A simple example might be a trigger handler interface and its associated base classes. These classes are reused together because they are tightly coupled to each other. Thus, they ought to be in the same package.
+
+#### The Common Reuse Principle (CRP)
+
+> **The components in a package should be closed together against the same kind of changes., A change that affects a package affects all components in that package**
+
+More important than reusability, is maintainability. When a change to components of an application is required, it would be beneficial if the changes were focused on a single package rather than across multiple packages. This ensures that a release for this change only contains the one package, other packages that don't depend upon the changed package do not need to be revalidated or rereleased.
```

Comment:
```suggestion
More important than reusability, is maintainability. When a change to components of an application is required, it would be beneficial if the changes were focused on a single package rather than across multiple packages. This ensures that a release for this change only contains the one package, other packages that don't depend upon the changed package do not need to be revalidated or re-released.
```